vlrouter.go

```diff
@@ -0,0 +1,662 @@
+		return err
+	}
+	err = portVlanFlow.Next(self.ipTable)
+	if err != nil {
+		log.Errorf("Error installing portvlan entry. Err: %v", err)
+		return err
+	}
+
+	// save the flow entry
+	self.portVlanFlowDb[endpoint.PortNo] = portVlanFlow
+
+	// Set the vlan and install it
+	// FIXME: Dont set the vlan till multi-vrf support. We cant pop vlan unless flow matches on vlan
+	// portVlanFlow.SetVlan(endpoint.Vlan)
+	outPort, err := self.ofSwitch.OutputPort(endpoint.PortNo)
+	if err != nil {
+		log.Errorf("Error creating output port %d. Err: %v", endpoint.PortNo, err)
+		return err
+	}
+
+	// Install the IP address
+	ipFlow, err := self.ipTable.NewFlow(ofctrl.FlowMatch{
+		Priority:  FLOW_MATCH_PRIORITY,
+		Ethertype: 0x0800,
+		IpDa:      &endpoint.IpAddr,
+	})
+
+	if err != nil {
+		log.Errorf("Error creating flow for endpoint: %+v. Err: %v", endpoint, err)
+		return err
+	}
+
+	destMacAddr, _ := net.ParseMAC(endpoint.MacAddrStr)
+
+	// Set Mac addresses
+	ipFlow.SetMacDa(destMacAddr)
+	ipFlow.SetMacSa(self.myRouterMac)
+
+	// Point the route at output port
+	err = ipFlow.Next(outPort)
+	if err != nil {
+		log.Errorf("Error installing flow for endpoint: %+v. Err: %v", endpoint, err)
+		return err
+	}
+
+	// Store the flow
+	self.flowDb[endpoint.IpAddr.String()] = ipFlow
+
+	if endpoint.EndpointType == "internal-bgp" {
+		return nil
+	}
+
+	//dial grpc server
+	conn, err := grpc.Dial("127.0.0.1:8080", grpc.WithBlock(), grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+
+	path := &api.Path{
+		Pattrs: make([][]byte, 0),
+	}
+
+	// form the path structure with appropriate path attributes
+	nlri := bgp.NewIPAddrPrefix(32, endpoint.IpAddr.String())
+	path.Nlri, _ = nlri.Serialize()
+	origin, _ := bgp.NewPathAttributeOrigin(bgp.BGP_ORIGIN_ATTR_TYPE_EGP).Serialize()
+	path.Pattrs = append(path.Pattrs, origin)
+	aspathParam := []bgp.AsPathParamInterface{bgp.NewAs4PathParam(2, []uint32{65002})}
+	aspath, _ := bgp.NewPathAttributeAsPath(aspathParam).Serialize()
+	path.Pattrs = append(path.Pattrs, aspath)
+	n, _ := bgp.NewPathAttributeNextHop(self.agent.routerIP).Serialize()
+	path.Pattrs = append(path.Pattrs, n)
+
+	name := ""
+
+	arg := &api.ModPathArguments{
+		Resource: api.Resource_GLOBAL,
+		Name:     name,
+		Paths:    []*api.Path{path},
+	}
+
+	//send arguement stream
+	client := api.NewGobgpApiClient(conn)
+	if client == nil {
+		log.Errorf("Gobgpapi stream invalid")
+		return nil
+	}
+
+	stream, err := client.ModPath(context.Background())
+	if err != nil {
+		log.Errorf("Fail to enforce Modpathi", err)
+		return err
+	}
+
+	err = stream.Send(arg)
+	if err != nil {
+		log.Errorf("Failed to send strean", err)
+		return err
+	}
+	stream.CloseSend()
+	res, e := stream.CloseAndRecv()
+	if e != nil {
+		log.Errorf("Falied toclose stream ")
+		return e
+	}
+	if res.Code != api.Error_SUCCESS {
+		return fmt.Errorf("error: code: %d, msg: %s", res.Code, res.Msg)
+	}
+
```

Comment:
you need to make self.policyAgent.AddEndpoint() call to add the endpoint to policy agent. This call sets up the dest-epg for the endpoint.